cmd2/parsing.py

```diff
@@ -25,91 +25,106 @@ class Statement(str):
-    :type output:           str or None
-    :var output_to:         if output was redirected, the destination file
-    :type output_to:        str or None
-
     """
     def __new__(cls,
-                obj: object,
-                *,
-                raw: str = None,
-                command: str = None,
-                args: str = None,
-                argv: List[str] = None,
-                multiline_command: str = None,
-                terminator: str = None,
-                suffix: str = None,
-                pipe_to: str = None,
-                output: str = None,
-                output_to: str = None
+                value: str,
+                raw: str = '',
+                command: str = '',
+                arg_list: List[str] = None,
+                multiline_command: str = '',
+                terminator: str = '',
+                suffix: str = '',
+                pipe_to: List[str] = None,
+                output: str = '',
+                output_to: str = ''
                 ):
+        """
+        :param value: The arguments, but not the command, nor the output redirection clauses.
+        :param raw: string containing exactly what we input by the user
+        :param command: the command, i.e. the first whitespace delimited word
+        :param arg_list: list of arguments to the command, not including any output
+                         redirection or terminators. quoted arguments remain quoted.
+        :param multiline_command: if the command is a multiline command, the name of the
+                                  command, otherwise empty
+        :param terminator: the character which terminated the multiline command, if
+                           there was one
+        :param suffix: characters appearing after the terminator but before output redirection, if any
+        :param pipe_to: if output was piped to a shell command, the shell command as a list of tokens
+        :param output: if output was redirected, the redirection token, i.e. '>>'
+        :param output_to: if output was redirected, the destination file
+        """
         """Create a new instance of Statement
 
         We must override __new__ because we are subclassing `str` which is
         immutable.
         """
-        stmt = str.__new__(cls, obj)
-        object.__setattr__(stmt, "raw", raw)
-        object.__setattr__(stmt, "command", command)
-        object.__setattr__(stmt, "args", args)
-        if argv is None:
-            argv = []
-        object.__setattr__(stmt, "argv", argv)
-        object.__setattr__(stmt, "multiline_command", multiline_command)
-        object.__setattr__(stmt, "terminator", terminator)
-        object.__setattr__(stmt, "suffix", suffix)
-        object.__setattr__(stmt, "pipe_to", pipe_to)
-        object.__setattr__(stmt, "output", output)
-        object.__setattr__(stmt, "output_to", output_to)
+        stmt = super().__new__(cls, value)
```

Comment:
You are taking a different and interesting approach to immutability which is subtly different than the original, but I think should behave identically in practice.

However, since ``cmd2`` already has a required dependency on ``attrs``, I would recommend considering using the ``attrs`` support for [immutability](http://www.attrs.org/en/stable/examples.html#immutability) for a couple reasons:
1. It is very succinct and readable
1. ``attrs`` is used by a ton of people, so it should be well tested and proven

Reply:
Agreed. I'll make the changes to `Statement` so it uses `attrs` for immutability.